### PR TITLE
repo-updater: add HTML version of repo-updater-state page

### LINKED ISSUE #8656: repo-updater: HTML version of repo-updater-state page
`repo-updater` currently exposes a useful debug endpoint for understanding what the git fetch scheduler is doing. However, the endpoint just dumps JSON. It would be great if instead we rendered a simple `html/template`:

- explain what each section is
- show human readable values of durations
- use html table

Example on our dogfood is at https://sourcegraph.sgdev.org/-/debug/repo-updater/repo-updater-state

Code is at https://github.com/sourcegraph/sourcegraph/blob/0c7c3e20efa5a65446fc1bb6a710c47fddf6ff10/cmd/repo-updater/shared/main.go#L217-L225

### PR DESCRIPTION
**Pull request by anukul**
Closes #8656

### CODE REVIEW

**Comment by keegancsmith** [comment]
Hi @anukul. Do you mind signing our CLA? https://forms.gle/FuroWPvTT8JfHiEE9

File: cmd/repo-updater/shared/assets/state.html.tmpl [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                    <tr>
+                        <td>{{.Repo.ID}}</td>
+                        <td>
+                            <a href="{{.Repo.URL}}"
+                               data-toggle="tooltip" data-placement="bottom" title="{{.Repo.URL}}">
+                                {{.Repo.Name}}
+                            </a>
+                        </td>
+                        {{/* truncate miliseconds using regex */}}
+                        <td>{{regexReplaceAll "\\..*s" (toString .Interval) "s"}}</td>
```
Comment (keegancsmith) [inline cmd/repo-updater/shared/assets/state.html.tmpl:48]: could you rather add a function which formats the `time.Duration` to the funcmap? Going via regex is less clear. Additionally it seems you introduced the `sprig` dependency for `regexReplaceAll`.

File: cmd/repo-updater/shared/main.go [hunk 2]
```diff
@@ -220,14 +222,13 @@ func Main(enterpriseInit EnterpriseInit) {
 			for _, dumper := range debugDumpers {
 				dumps = append(dumps, dumper.DebugDump())
 			}
-
-			p, err := json.MarshalIndent(dumps, "", "  ")
```
Comment (keegancsmith) [inline cmd/repo-updater/shared/main.go:224]: the json data is still useful for data analysis. Can you check the accept header to do content negotiation here. We already use `github.com/golang/gddo/httputil.NegotiateContentType` elsewhere in our codebase.

File: cmd/repo-updater/shared/main.go [hunk 3]
```diff
@@ -220,14 +222,13 @@ func Main(enterpriseInit EnterpriseInit) {
 			for _, dumper := range debugDumpers {
 				dumps = append(dumps, dumper.DebugDump())
```
Comment (keegancsmith) [inline cmd/repo-updater/shared/main.go:223]: notice we can have other `DebugDump` here. The HTML template you created only assumes the `scheduler.DebugDump`. I see you do `first .` on it, making it clearer you got rid of the rest.

I suggest maybe just a textarea for the rest of the dumps? Or maybe a link to the raw dump (eg check for `?accept=application/json` then you can link to it directly).

Comment (anukul) [reply cmd/repo-updater/shared/main.go:223]: oops :s

I've now added an alert that shows up when there is more than one `DebugDump`.  
Clicking on this takes the user to `?format=json` (picked up from [ui/raw.go](https://github.com/sourcegraph/sourcegraph/blob/3.18/cmd/frontend/internal/app/ui/raw.go#L113-L115))

image

File: cmd/repo-updater/shared/assets/state.html.tmpl (hunk 1 above)
Comment (anukul) [reply cmd/repo-updater/shared/assets/state.html.tmpl]: sure.
btw I ran `go mod tidy` to remove `sprig`, and it also removed a bunch of unrelated stuff.
please let me know if that should not be the case, i can remove `sprig` manually.

**Comment by tsenart** [comment]
Changelog entry?

**Comment by anukul** [comment]
@tsenart @keegancsmith Should I add that in a new PR?

**Comment by tsenart** [comment]
That'd work!

**Comment by unknwon** [comment]
Thanks for the awesome addition to this page! ❤️ 

However, this page also contains update queue dump of permissions (you can find it when request JSON format), but I don't see it's been reflected in the fancy HTML version? Could be an oversight 🤔

**Comment by anukul** [comment]
@unknwon Oh.. looks like my local env was not configured to run [`enterpriseInit`](https://github.com/sourcegraph/sourcegraph/blob/3.18/enterprise/cmd/repo-updater/main.go#L39) and consequently the permissions syncing manager. I'll try to set it up and fix this.

**Comment by keegancsmith** [comment]
@unknwon see this comment above https://github.com/sourcegraph/sourcegraph/pull/12498#discussion_r461156758

You can still access it via the raw json dump.

**Pull request merged**